### PR TITLE
add a doc for prefetchfiles-nri-plugin

### PR DESCRIPTION
**Pull request by billie60**
This document mainly introduces the early configuration and usage of prefetch-nri-plugin.

### CODE REVIEW

File: docs/prefetchfiles-nri-plugin.md [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+To improve the flexibility of nydus image files prefetch, for the k8s scenario, we can specify a prefetch files list when create a nydus daemon. The prefetch files list is user self-defined. Nydus-snapshotter has implemented a containerd NRI plugin to transmit the path of prefetch files list to nydus-snapshotter. The prefetch plugin requires NRI 2.0, which is available in containerd (>=v1.7.0). The prefetch plugin subscribes pod creation event, fetches the path of prefetch files list and forwards it to nydus-snapshotter. Then when `nydusd` starts, it will pull the files defined in the prefetch files list through lazy loading. This allows the pull of the prefetch files to be done during container creation rather than image convert, improving the flexibility of file prefetching.
+
+## Requirements
+
+- [NRI 2.0](https://github.com/containerd/nri): Which has been integrated into containerd since [v1.7.0](https://github.com/containerd/containerd/tree/v1.7.0-beta.1).
+
+## Workflow
+
+1. Add information such as image reference and prefetch list path to annotations in pod configuration file.
+2. Run the prefetch plugin to monitoring RunPodSandbox events.
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:11]: Bad syntax, should be "to monitor"

File: docs/prefetchfiles-nri-plugin.md [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+3. The prefetch plugin fetches image reference and prefetch files path and forwards them to nydus-snapshotter.
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:12]: file paths

File: docs/prefetchfiles-nri-plugin.md [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+4. Nydus-snapshotter specifies the prefetch list when starting nydus daemon.
+5. Nydusd completes the mounting of the nydus image.
+<!-- 4. Run the prefetch plugin to fetch those information and forward them to nydus-snapshotter when the RunPodSandbox event occurs.  -->
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:15]: What is this line for?

File: docs/prefetchfiles-nri-plugin.md [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+      command: ["ls"]
+    - name: container-3
+      image: dockerhub.kubekey.local/dfns/wordpress:nydus_latest
+      command: ["ls"]
+EOF
+```
+
+
+
+Note that the naming of keys in annotations is fixed, and the values in annotations is user self-defined.
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:132]: are user self-defined.

File: docs/prefetchfiles-nri-plugin.md [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+After create a pod, image reference and path of prefetch list will received by nydus-snapshotter.
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:133]: After creating a pod

File: docs/prefetchfiles-nri-plugin.md [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+metadata:
+  name: wordpress-sandbox
+  namespace: default
+  attempt: 1
+  uid: hdishd83djaidwnduwk28bcsb
+log_directory: /tmp
+annotations:
+ containerd.io/nydus-prefetch: |
+    [
+    {"image": "dockerhub.kubekey.local/dfns/wordpress:nydus_latest", "prefetch": "/path/to/wordpress_prefetch.txt"}
```
Comment (changweige) [inline docs/prefetchfiles-nri-plugin.md:70]: Um. If the value is a file path, I am concerning how can we distribute the file to each K8s node?

File: docs/prefetchfiles-nri-plugin.md [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+```
+
+Restart the containerd service.
+
+```console
+sudo systemctl restart containerd
+```
+Add a prefetch configuration file to specify the correct socket address. Note that this configuration file is designed for starting the NRI plugin in `pre-connection` mode. In this mode, the NRI plugin and containerd will start simultaneously.
+```console
+sudo tee /etc/nydus/prefetchConfig.toml <<- EOF
```
Comment (sctb512) [inline docs/prefetchfiles-nri-plugin.md:48]: Is `/etc/nydus/prefetchConfig.toml` the right path? In default, containerd read NRI's configuration file from `/etc/nri/conf.d`.

File: docs/prefetchfiles-nri-plugin.md (hunk 6 above)
Comment (sctb512) [reply docs/prefetchfiles-nri-plugin.md:70]: Maybe add an comment: specify the `prefetch` value with an URL is coming. 😉

File: docs/prefetchfiles-nri-plugin.md [hunk 8]
```diff
@@ -0,0 +1,156 @@
+# User self-defined nydus image files prefetch
+To improve the flexibility of nydus image files prefetch, for the k8s scenario, we can specify a prefetch files list when create a nydus daemon. The prefetch files list is user self-defined. Nydus-snapshotter has implemented a containerd NRI plugin to transmit the path of prefetch files list to nydus-snapshotter. The prefetch plugin requires NRI 2.0, which is available in containerd (>=v1.7.0). The prefetch plugin subscribes pod creation event, fetches the path of prefetch files list and forwards it to nydus-snapshotter. Then when `nydusd` starts, it will pull the files defined in the prefetch files list through lazy loading. This allows the pull of the prefetch files to be done during container creation rather than image convert, improving the flexibility of file prefetching.
```
[the remark this answers is not in the document]
Comment (billie60) [reply docs/prefetchfiles-nri-plugin.md:2]: Thanks for your comment. What is saved in the prefetch file is indeed the list of prefetch file paths.  But the --prefetch-files parameter of nydusd needs to specify a path that points to the file containing the prefetch files (maybe a txt file). So I mean the path of prefetch files list, which is the actual parameter of `--prefetch-files`

File: docs/prefetchfiles-nri-plugin.md (hunk 6 above)
Comment (billie60) [reply docs/prefetchfiles-nri-plugin.md:69]: THANKS very much! Other comments have been modified, and I will also correct this comment in the next version commit

**Comment by imeoer** [comment]
The `.idea` directory should be removed.

File: integration/prefetch.sh
Comment (sctb512) [inline integration/prefetch.sh:1]: Is it necessary to create a new script for integration?
Morever, `entrypoint.sh` is used as the `ENTRYPOINT` of image, but this script `integration/prefetch.sh` is not used by `make integrate` command.

File: pkg/prefetch/prefetch.go [hunk 9]
```diff
@@ -7,12 +7,22 @@
...
+	"net/http"
+	"os"
+	"strings"
 	"sync"
 
 	"github.com/containerd/containerd/log"
+	"golang.org/x/net/html"
 )
 
+const prefetchlistPath string = "/opt/PrefetchFiles"
```
Comment (sctb512) [inline pkg/prefetch/prefetch.go:24]: Is this configurable?

File: pkg/prefetch/prefetch.go [hunk 10]
```diff
@@ -34,14 +44,88 @@ func (p *prefetchInfo) SetPrefetchFiles(body []byte) error {
...
+// Obtain the prefetch files through the URL and transfer them to a local txt file.
+func (p *prefetchInfo) getprefetchConfigFile(url, image string) (string, error) {
+	resp, err := http.Get(url)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		errMsg := fmt.Sprintf("HTTP response status code is not 200 OK: %s", resp.Status)
```
Comment (sctb512) [inline pkg/prefetch/prefetch.go:68]: `"failed to GET %s, response status $v", url, resp.Status`

File: pkg/prefetch/prefetch.go [hunk 11]
```diff
@@ -34,14 +44,88 @@ func (p *prefetchInfo) SetPrefetchFiles(body []byte) error {
...
+		if err = file.Close(); err != nil {
+			log.L.Fatalf("Error while closing the file: %v", err)
+		}
+	}()
+	var htmlBuffer bytes.Buffer
+
+	tokenizer := html.NewTokenizer(resp.Body)
+	inBody := false
+
+	for {
```
Comment (sctb512) [inline pkg/prefetch/prefetch.go:95]: I don't understand why we need this loop. May have a more concise implementation? 
Could you take a look at this? cc @imeoer

**Review by sctb512 — COMMENTED** [review]
Others, LGTM

File: pkg/prefetch/prefetch.go (hunk 11 above)
Comment (imeoer) [reply pkg/prefetch/prefetch.go:95]: Yes, we can't assume the server side returns such html format.

File: Makefile [hunk 12]
```diff
@@ -135,6 +135,7 @@ smoke:
 integration:
 	CGO_ENABLED=1 ${PROXY} GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X "${PKG}/version.Version=${VERSION}" -extldflags "-static"' -race -v -o bin/containerd-nydus-grpc ./cmd/containerd-nydus-grpc
 	CGO_ENABLED=1 ${PROXY} GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X "${PKG}/version.Version=${VERSION}" -extldflags "-static"' -race -v -o bin/nydus-overlayfs ./cmd/nydus-overlayfs
+	CGO_ENABLED=1 ${PROXY} GOOS=${GOOS} GOARCH=${GOARCH} go build -ldflags '-X "${PKG}/version.Version=${VERSION}" -extldflags "-static"' -race -v -o bin/03-prefetchfiles-nri-plugin ./cmd/prefetchfiles-nri-plugin
```
Comment (imeoer) [inline Makefile:138]: Why need `03-` prefix ?

File: config/config.go [hunk 13]
```diff
@@ -208,6 +208,9 @@ type SystemControllerConfig struct {
 	Address     string      `toml:"address"`
 	DebugConfig DebugConfig `toml:"debug"`
 }
+type PrefetchConfig struct {
```
Comment (imeoer) [inline config/config.go:211]: Should add space before this.

File: docs/prefetchfiles-nri-plugin.md [hunk 14]
```diff
@@ -0,0 +1,127 @@
+# User self-defined nydus image files prefetch
+To improve the flexibility of nydus image files prefetch, for the k8s scenario, we can specify a prefetch files list when create a nydus daemon. The prefetch files list is user self-defined. Nydus-snapshotter has implemented a containerd NRI plugin to transmit the path of prefetch files list to nydus-snapshotter. The prefetch plugin requires NRI 2.0, which is available in containerd (>=v1.7.0). The prefetch plugin subscribes pod creation event, obtains the URL address containing the content of the files need to be prefetched, and forwards it to `nydus-snapshotter`. The `nydus-snapshotter` reads the data through the URL and stores it locally. Then when `nydusd` starts, it will pull the files defined in the prefetch files list through lazy loading. This allows the pull of the prefetch files to be done during container creation rather than image convert, improving the flexibility of file prefetching.
```
Comment (imeoer) [inline docs/prefetchfiles-nri-plugin.md:2]: First should explain why we need the prefetch, then say the flexibility of prefetch on runtime.

File: docs/prefetchfiles-nri-plugin.md [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+
+## Requirements
+
+- [NRI 2.0](https://github.com/containerd/nri): Which has been integrated into containerd since [v1.7.0](https://github.com/containerd/containerd/tree/v1.7.0-beta.1).
```
Comment (imeoer) [inline docs/prefetchfiles-nri-plugin.md:6]: Should be `https://github.com/containerd/containerd/tree/v1.7.0` ?

File: docs/prefetchfiles-nri-plugin.md [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+
+1. Add information such as image reference and URL address containing prefetch files to annotations in pod configuration file.
+2. Run the prefetch plugin to monitor RunPodSandbox events.
+3. The prefetch plugin fetches image reference and URL and forwards them to nydus-snapshotter.
+4. Nydus-snapshotter specifies the prefetch list when starting nydus daemon.
+5. Nydusd completes the mounting of the nydus image.
+
+
+
+
```
Comment (imeoer) [inline docs/prefetchfiles-nri-plugin.md:18]: Remove multiple spaces before this.

File: docs/prefetchfiles-nri-plugin.md [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+## Modify configuration file 
+
+Modify containerd's toml configuration file to enable NRI.
+```console
+sudo tee -a /etc/containerd/config.toml <<- EOF
+[plugins."io.containerd.nri.v1.nri"]
```
Comment (imeoer) [inline docs/prefetchfiles-nri-plugin.md:25]: Does this config will conflict with optimizer plugin?

File: docs/prefetchfiles-nri-plugin.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+kind: pod
+metadata:
+  name: wordpress-sandbox
+  namespace: default
+  attempt: 1
+  uid: hdishd83djaidwnduwk28bcsb
+log_directory: /tmp
+annotations:
+ containerd.io/nydus-prefetch: |
+    [
```
Comment (imeoer) [inline docs/prefetchfiles-nri-plugin.md:76]: Format it:

```
[
    {
        "image": "ghcr.io/dragonflyoss/image-service/wordpress:nydus-nightly-v6",
        "prefetch": "http://example.com/api/v1/resource/wordpress",
    }
]
```

Comment (imeoer) [reply docs/prefetchfiles-nri-plugin.md:76]: Does the annotation work with multiple line value?

File: go.mod [hunk 19]
```diff
@@ -54,6 +54,17 @@ require (
...
 )
 
+require (
+	github.com/cilium/ebpf v0.9.1 // indirect
+	github.com/coreos/go-systemd/v22 v22.5.0 // indirect
+	github.com/docker/go-units v0.5.0 // indirect
+	github.com/godbus/dbus/v5 v5.1.0 // indirect
+)
+
+require github.com/freddierice/go-losetup v0.0.0-20220711213114-2a14873012db
```
Comment (imeoer) [inline go.mod:64]: Please merge `require` section.

File: misc/prefetch/crictl.yaml [hunk 20]
```diff
@@ -0,0 +1,4 @@
+runtime-endpoint: unix:///run/containerd/containerd.sock
+image-endpoint: unix:///run/containerd/containerd.sock
+timeout: 10
+debug: true
```
Comment (imeoer) [inline misc/prefetch/crictl.yaml:4]: Add blank line here.

File: misc/prefetch/prefetch_server.go [hunk 21]
```diff
@@ -0,0 +1,94 @@
+package main
+
+import (
+	"fmt"
+	"net/http"
+)
+
+func wordpress_prefetchlist(w http.ResponseWriter, req *http.Request) {
```
Comment (imeoer) [inline misc/prefetch/prefetch_server.go:8]: `getWordpressPrefetchList`

File: misc/prefetch/prefetch_server.go [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+/usr/src/wordpress/wp-includes/js/customize-preview-widgets.min.js
+/usr/src/wordpress/wp-includes/js/json2.js
+/usr/src/wordpress/wp-includes/js/wp-ajax-response.js
+/usr/src/wordpress/wp-includes/js/zxcvbn-async.js
+/usr/src/wordpress/wp-includes/js/tinymce/license.txt`
+
+	fmt.Fprintf(w, responseText)
+}
+
+func tomcat_prefetchlist(w http.ResponseWriter, req *http.Request) {
```
Comment (imeoer) [inline misc/prefetch/prefetch_server.go:40]: `getTomcatPrefetchList`

File: misc/snapshotter/config.toml [hunk 23]
```diff
@@ -63,6 +63,9 @@ address = ":9110"
 [remote]
 convert_vpc_registry = false
 
+[prefetch]
+dir = "/opt/prefetch"
```
Comment (imeoer) [inline misc/snapshotter/config.toml:67]: How do we remove the useless prefetch list file here?

File: pkg/prefetch/prefetch.go [hunk 24]
```diff
@@ -8,16 +8,28 @@ package prefetch
...
 
 type prefetchInfo struct {
 	prefetchMap   map[string]string
 	prefetchMutex sync.Mutex
 }
 
+type prefetchlist struct {
+	PrefetchFiles []string `json:"prefetchFiles"`
+}
+
```
Comment (imeoer) [inline pkg/prefetch/prefetch.go:32]: type prefetchList struct {
	Files []string `json:"files"`
}

File: pkg/prefetch/prefetch.go [hunk 25]
```diff
@@ -34,14 +46,87 @@ func (p *prefetchInfo) SetPrefetchFiles(body []byte) error {
...
+		}
 		p.prefetchMap[image] = prefetchfiles
 	}
 
 	log.L.Infof("received prefetch list from nri plugin: %v ", p.prefetchMap)
 	return nil
 }
 
+// Obtain the prefetch files through the URL and transfer them to local files().
+func (p *prefetchInfo) getprefetchConfigFile(url, image string) (string, error) {
```
Comment (imeoer) [inline pkg/prefetch/prefetch.go:62]: getPrefetchListPath

File: docs/prefetchfiles-nri-plugin.md (hunk 18 above)
Comment (billie60) [reply docs/prefetchfiles-nri-plugin.md:79]: yes, if there are multiple images and prefetchlists in annotations, the code works correctly.

**Pull request closed without merging**